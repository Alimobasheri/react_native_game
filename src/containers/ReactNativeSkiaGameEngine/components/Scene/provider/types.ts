/**
 * Props for the SceneProvider component.
 */
export interface ISceneProviderProps {
  /**
   * Name of the default scene.
   */
  defaultSceneName: string;
}
