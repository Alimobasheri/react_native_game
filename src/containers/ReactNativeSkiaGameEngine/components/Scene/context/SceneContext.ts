import { Camera } from '@/containers/ReactNativeSkiaGameEngine/types';
import { createContext } from 'react';
import { ISceneTransitionState } from '../types/transitions';
import { SharedValue } from 'react-native-reanimated';

/**
 * Context value for the Scene component.
 */
export interface IScenContextValue {
  /**
   * Current active scenes.
   */
  activeScenes: Record<string, boolean>;
  /**
   * Function to enable a scene.
   * @param name - Name of the scene.
   * @returns {void}
   */
  enableScene: (name: string) => void;
  /**
   * Function to disable a scene.
   * @param name - Name of the scene.
   * @returns {void}
   */
  disableScene: (name: string) => void;
  /**
   * Function to switch to a scene.
   * @param name - Name of the scene.
   * @param options - Options for preserving the history.
   * @returns {void}
   */
  switchScene: (name: string, options?: { preserveHistory: boolean }) => void;
  /**
   * Function to go back to the previous scene.
   * @returns {void}
   */
  goBack: () => void;
  /**
   * Function to register a scene.
   * @param name - Name of the scene.
   * @param isActive - Whether the scene is active.
   * @returns {void}
   */
  registerScene: (name: string, isActive: boolean) => void;
  /**
   * The anme of the scene that is rendering this sceneProvider
   **/
  name: string;
  /**
   * Reference Of the camera registered for the current Scene.
   */
  sceneCamera: Camera | null;
  /**
   * A sharedValue including lates Transition Phase, Transition Progress and A Reference Of the camera registered for the current Scene.
   */
  sceneTransitionState: SharedValue<ISceneTransitionState> | null;
}

export const SceneContext = createContext<IScenContextValue | null>(null);
