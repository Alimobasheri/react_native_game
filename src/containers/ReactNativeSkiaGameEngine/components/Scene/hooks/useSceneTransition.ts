import { useAnimationsController } from '@/containers/ReactNativeSkiaGameEngine/hooks/useAnimationsController/useAnimationsController';
import { ActiveAnimation } from '@/containers/ReactNativeSkiaGameEngine/services/Animations';
import { Camera } from '@/containers/ReactNativeSkiaGameEngine/types';
import { createTimingAnimation } from '@/containers/ReactNativeSkiaGameEngine/utils';
import { useEffect, useRef, useState } from 'react';
import {
  useSharedValue,
  runOnJS,
  useAnimatedReaction,
  SharedValue,
  runOnUI,
} from 'react-native-reanimated';
import { SceneTransition, TransitionPhase } from '../types/transitions';

interface ITransitionConfig {
  duration?: number;
  enterDuration?: number;
  exitDuration?: number;
}

interface IUseSceneTransitionProps {
  isActive: boolean;
  camera: Camera;
  enter?: SceneTransition | null;
  exit?: SceneTransition | null;
  config: ITransitionConfig;
}

/**
 * Hook that manages scene transitions (enter and exit animations) for a scene.
 * It provides animation properties that can be applied to the scene's root component.
 *
 * @param {boolean} isActive - Whether the scene is currently active.
 * @param {SceneTransition} enter - The transition callback for enter phase.
 * @param {SceneTransition} exit - The transition callbakc for exit phase.
 * @param {Object} config - Configuration for transition durations.
 * @param {number} [config.duration=500] - Default duration for transitions.
 * @param {number} [config.enterDuration] - Specific duration for the enter transition (overrides `duration`).
 * @param {number} [config.exitDuration] - Specific duration for the exit transition (overrides `duration`).
 *
 * @returns {object} returnObject
 * @returns {boolean} returnObject.isTransitioning - Whether the scene is currently transitioning.
 *
 * @example
 * const { props } = useSceneTransition(true, 'fade', 'slide', { duration: 300 });
 * <Rect {...props}>Scene Content</Rect>
 */
export const useSceneTransition = ({
  isActive,
  camera,
  enter = null,
  exit = null,
  config = { duration: 500 },
}: IUseSceneTransitionProps) => {
  const enterDuration = config.enterDuration ?? config.duration ?? 500;
  const exitDuration = config.exitDuration ?? config.duration ?? 500;
  const duration = isActive ? enterDuration : exitDuration;

  const { registerAnimation, removeAnimation } = useAnimationsController();

  let animation = useRef<ActiveAnimation | null>(null);

  const [isTransitioning, setIsTransitioning] = useState(false);

  const phase = useSharedValue(
    isActive ? TransitionPhase.BeforeEnter : TransitionPhase.Idle
  );
  const progress = useSharedValue(isActive ? 0 : 1);

  const isInitialRender = useRef<boolean>(true);

  useEffect(() => {
    if (duration === 0) {
      progress.value = isActive ? 1 : 0;
    } else if (!isInitialRender.current || isActive) {
      if (isActive && phase.value === TransitionPhase.Idle) {
        phase.value = TransitionPhase.BeforeEnter;
      }
      if (isActive && phase.value === TransitionPhase.BeforeEnter) {
        if (typeof enter === 'function') {
          runOnUI(enter)({ camera, phase: phase.value, progress });
        }
        phase.value = TransitionPhase.Enter;
      } else if (!isActive && phase.value === TransitionPhase.AfterEnter) {
        phase.value = TransitionPhase.BeforeExit;
        if (typeof exit === 'function') {
          runOnUI(exit)({ camera, phase: phase.value, progress });
        }
        phase.value = TransitionPhase.Exit;
      }
      setIsTransitioning(true);
      animation.current = registerAnimation(
        progress,
        createTimingAnimation(progress.value, isActive ? 1 : 0, duration),
        {
          duration,
          removeOnComplete: true,
          onDone: () => {
            runOnJS(setIsTransitioning)(false);
          },
        }
      );
    }

    isInitialRender.current = false;

    return () => {
      if (animation.current) {
        removeAnimation(animation.current);
      }
    };
  }, [isActive]);

  useAnimatedReaction(
    () => {
      return progress;
    },
    (progress) => {
      console.log('ðŸš€ ~ progress:', progress.value, phase.value);
      if (
        typeof enter === 'function' &&
        phase.value === TransitionPhase.Enter
      ) {
        enter({ camera, phase: phase.value, progress });
      } else if (
        typeof exit === 'function' &&
        phase.value === TransitionPhase.Exit
      ) {
        exit({ camera, phase: phase.value, progress });
      }
      if (isActive && progress.value === 1)
        phase.value = TransitionPhase.AfterEnter;
      else if (!isActive && progress.value === 0)
        phase.value = TransitionPhase.Idle;
    }
  );

  return {
    isTransitioning,
  };
};
