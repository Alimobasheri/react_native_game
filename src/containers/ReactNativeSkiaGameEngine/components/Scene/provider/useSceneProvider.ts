import { Camera } from '@/containers/ReactNativeSkiaGameEngine/types';
import { useCallback, useEffect, useState } from 'react';

export interface IUseSceneProviderArgs {
  camera?: Camera;
}

/**
 * Hook that provides scene management functionality, such as enabling/disabling scenes,
 * switching between them, and handling scene history.
 *
 * @returns {Object} sceneContext - The context value to be provided by the `SceneProvider`.
 * @returns {Object} sceneContext.activeScenes - The currently active scenes.
 * @returns {Function} sceneContext.enableScene - Function to activate a scene.
 * @returns {Function} sceneContext.disableScene - Function to deactivate a scene.
 * @returns {Function} sceneContext.switchScene - Switches between active scenes.
 * @returns {Function} sceneContext.goBack - Navigates to the previous scene in history.
 * @returns {Function} sceneContext.registerScene - Registers a new scene.
 * @returns {Function} sceneContext.pushScene - Adds a scene to the history stack.
 * @returns {Function} sceneContext.replaceScene - Replaces the current scene in the history stack.
 *
 * @example
 * const { activeScenes, enableScene, disableScene } = useSceneProvider();
 * enableScene('level1');
 * disableScene('level2');
 */
export const useSceneProvider = ({ camera }: IUseSceneProviderArgs) => {
  const [activeScenes, setActiveScenes] = useState<Record<string, boolean>>({});
  const [sceneHistory, setSceneHistory] = useState<string[]>([]);
  const [sceneCamera, setSceneCamera] = useState<Camera | null>(camera || null);

  const enableScene = useCallback(
    (name: string) => setActiveScenes((prev) => ({ ...prev, [name]: true })),
    [setActiveScenes]
  );
  const disableScene = useCallback(
    (name: string) => setActiveScenes((prev) => ({ ...prev, [name]: false })),
    [setActiveScenes]
  );

  const switchScene = useCallback(
    (sceneName: string) => {
      setActiveScenes((prevScenes) => {
        const updatedScenes = Object.keys(prevScenes).reduce((acc, key) => {
          acc[key] = false;
          return acc;
        }, {} as Record<string, boolean>);

        updatedScenes[sceneName] = true;

        return updatedScenes;
      });
    },
    [setActiveScenes]
  );

  const goBack = useCallback(() => {
    if (sceneHistory.length > 1) {
      const previousScene = sceneHistory[sceneHistory.length - 2];
      switchScene(previousScene);
      setSceneHistory(sceneHistory.slice(0, -1));
    }
  }, [sceneHistory, enableScene, disableScene]);

  const pushScene = useCallback((sceneName: string) => {
    switchScene(sceneName);
    setSceneHistory((prev) => [...prev, sceneName]);
  }, []);

  const replaceScene = useCallback((sceneName: string) => {
    switchScene(sceneName);
    setSceneHistory((sceneHistory) =>
      sceneHistory.slice(0, -1).concat([sceneName])
    );
  }, []);

  const registerScene = useCallback((name: string, isActive: boolean) => {
    setActiveScenes((prev) => ({
      ...prev,
      [name]: isActive,
    }));

    if (isActive) {
      pushScene(name);
    }
  }, []);

  useEffect(() => {
    setSceneCamera(camera || null);
  }, [sceneCamera]);

  return {
    activeScenes,
    sceneCamera,
    enableScene,
    disableScene,
    switchScene,
    goBack,
    registerScene,
    pushScene,
    replaceScene,
  };
};
