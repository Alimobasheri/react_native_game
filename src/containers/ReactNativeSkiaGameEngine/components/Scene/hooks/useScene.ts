import { useContext } from 'react';
import { SceneContext } from '../context';

export const useScene = () => {
  const seceneContext = useContext(SceneContext);
  if (!seceneContext) {
    throw new Error('useScene must be used within a SceneProvider');
  }
  const { enableScene, disableScene, activeScenes } = seceneContext;
  return { enableScene, disableScene, activeScenes };
};
