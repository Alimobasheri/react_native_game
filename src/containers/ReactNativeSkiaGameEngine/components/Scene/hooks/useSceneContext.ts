import { useContext, useMemo } from 'react';
import { SceneContext } from '../context';
import { ISceneProps } from '../Scene';

export const useSceneContext = ({ defaultSceneName }: ISceneProps) => {
  const seceneContext = useContext(SceneContext);
  if (!seceneContext) {
    throw new Error('useSceneContext must be used within a SceneProvider');
  }
  const { activeScenes } = seceneContext;
  const isActive = useMemo(
    () => activeScenes[defaultSceneName] ?? false,
    [activeScenes?.[defaultSceneName]]
  );

  return { isActive };
};
