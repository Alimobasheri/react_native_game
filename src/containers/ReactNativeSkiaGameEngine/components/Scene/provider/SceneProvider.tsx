import { FC, PropsWithChildren } from 'react';
import { SceneContext } from '../context';
import { useSceneProvider } from './useSceneProvider';
import { Camera } from '@/containers/ReactNativeSkiaGameEngine/types';
import { SharedValue } from 'react-native-reanimated';
import { ISceneTransitionState } from '../types/transitions';

export interface ISceneProviderProps {
  camera?: Camera;
  sceneTransitionState?: SharedValue<ISceneTransitionState>;
}

/**
 * The `SceneProvider` component wraps scenes and provides the scene context
 * to handle active scene management (enabling, disabling, transitioning scenes).
 *
 * @component
 * @example
 * <SceneProvider>
 *   <Scene defaultSceneName="gameScene" isActive={true}>
 *     <Text>Game Scene</Text>
 *   </Scene>
 * </SceneProvider>
 *
 * @param {React.ReactNode} children - The scene components to be rendered within the provider.
 * @returns {JSX.Element} Returns the wrapped scene with context.
 */
export const SceneProvider: FC<PropsWithChildren<ISceneProviderProps>> = ({
  camera,
  sceneTransitionState,
  children,
}) => {
  const value = useSceneProvider({ camera, sceneTransitionState });
  return (
    <SceneContext.Provider value={value}>{children}</SceneContext.Provider>
  );
};
