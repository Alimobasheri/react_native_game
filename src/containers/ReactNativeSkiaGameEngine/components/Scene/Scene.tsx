import React, { FC, PropsWithChildren } from 'react';
import { useSceneContext } from './hooks/useSceneContext';
import { useSceneTransition } from './hooks/useSceneTransition';
import { SceneProvider } from './provider';

export interface ISceneProps extends PropsWithChildren {
  defaultSceneName: string;
  enter?: 'fade' | 'slide' | 'zoom' | null;
  exit?: 'fade' | 'slide' | 'zoom' | null;
  transitionConfig?: {
    duration?: number;
    enterDuration?: number;
    exitDuration?: number;
  };
  x?: number;
  y?: number;
  width?: number;
  height?: number;
  rootComponent?: React.ComponentType<any>;
  rootComponentProps?: Record<string, any>;
  isActive?: boolean;
}

/**
 * `Scene` component handles individual game scenes within the game engine.
 * It supports custom transitions (fade, slide, zoom) when entering or exiting scenes.
 * Scenes can be nested inside one another, and the visibility of each scene is controlled
 * by the `isActive` prop and the scene context. Nested scenes will be rendered only if both
 * the parent and the nested scenes are active.
 *
 * @component
 * @example
 * // A basic scene with a fade transition
 * <Scene
 *   defaultSceneName="gameOver"
 *   enter="fade"
 *   exit="fade"
 *   isActive={true}
 *   transitionConfig={{ duration: 300 }}
 * >
 *   <Text>Game Over</Text>
 * </Scene>
 *
 * @example
 * // Nesting scenes where both parent and child scenes are active
 * <Scene
 *   defaultSceneName="parentScene"
 *   enter="slide"
 *   exit="fade"
 *   isActive={true}
 * >
 *   <Scene
 *     defaultSceneName="childScene"
 *     enter="zoom"
 *     exit="slide"
 *     isActive={true}
 *   >
 *     <Text>Child Scene Content</Text>
 *   </Scene>
 *   <Text>Parent Scene Content</Text>
 * </Scene>
 *
 * @param {Object} props - The properties object.
 * @param {string} props.defaultSceneName - The name of the scene (used for identification).
 * @param {'fade'|'slide'|'zoom'|null} [props.enter=null] - The animation to use when the scene enters.
 * @param {'fade'|'slide'|'zoom'|null} [props.exit=null] - The animation to use when the scene exits.
 * @param {Object} [props.transitionConfig={ duration: 500 }] - Transition configuration for enter/exit animations.
 * @param {number} [props.transitionConfig.duration=500] - Duration for both enter and exit transitions.
 * @param {number} [props.transitionConfig.enterDuration] - Specific duration for the enter transition (overrides `duration`).
 * @param {number} [props.transitionConfig.exitDuration] - Specific duration for the exit transition (overrides `duration`).
 * @param {number} [props.x=0] - X-coordinate for scene positioning.
 * @param {number} [props.y=0] - Y-coordinate for scene positioning.
 * @param {number} [props.width=300] - Width of the scene.
 * @param {number} [props.height=300] - Height of the scene.
 * @param {React.ComponentType} [props.rootComponent='Rect'] - Component to be used as the root of the scene (default is Rect).
 * @param {Object} [props.rootComponentProps={}] - Props to pass to the root component.
 * @param {boolean} [props.isActive=false] - Whether the scene is active (i.e., visible). If `false`, the scene and its children will not be rendered.
 *
 * @returns {JSX.Element|null} - Returns the JSX for the scene if active, otherwise returns null.
 *
 * @see {@link useSceneContext} for handling scene activity.
 * @see {@link useSceneTransition} for handling scene transitions.
 */
export const Scene: FC<ISceneProps> = ({
  defaultSceneName,
  children,
  enter = null,
  exit = null,
  transitionConfig = { duration: 500 },
  x = 0,
  y = 0,
  width = 300,
  height = 300,
  rootComponent: RootComponent = 'Rect',
  rootComponentProps = {},
  isActive = false,
}) => {
  const { isActive: currentIsActive } = useSceneContext(
    defaultSceneName,
    isActive
  );

  const { props } = useSceneTransition(
    currentIsActive,
    enter,
    exit,
    transitionConfig
  );

  return currentIsActive ? (
    <SceneProvider>
      <RootComponent
        x={x}
        y={y}
        width={width}
        height={height}
        {...rootComponentProps}
        {...props}
      >
        {children}
      </RootComponent>
    </SceneProvider>
  ) : null;
};
