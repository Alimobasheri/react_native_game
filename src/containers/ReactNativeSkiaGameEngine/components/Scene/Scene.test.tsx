import React from 'react';
import { render } from '@testing-library/react-native';
import { Scene } from './Scene';
import { SceneContext } from './context/SceneContext';
import { SceneProvider } from './provider/SceneProvider';
import { useSceneContext } from './hooks/useSceneContext';
import { Text } from 'react-native';

jest.mock('./provider/SceneProvider');
jest.mock('./hooks/useSceneContext');

describe('Scene', () => {
  const mockSceneProvider = jest.fn();
  const mockUseSceneContext = jest.fn();

  const mockContextValue = {
    activeScenes: { testScene: true },
    enableScene: jest.fn(),
    disableScene: jest.fn(),
  };

  beforeEach(() => {
    (SceneProvider as jest.Mock).mockImplementation(({ children }) => (
      <>{children}</>
    ));
    mockUseSceneContext.mockReturnValue({ isActive: true });
    (useSceneContext as jest.Mock).mockImplementation(mockUseSceneContext);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const RenderWithContext = (props = {}) => (
    <SceneContext.Provider value={mockContextValue}>
      <Scene defaultSceneName="testScene" {...props}>
        <Text>Test Child Component</Text>
      </Scene>
    </SceneContext.Provider>
  );

  test('should render the SceneProvider if the scene is active', () => {
    const { getByText } = render(<RenderWithContext />);

    expect(SceneProvider).toHaveBeenCalledWith(
      expect.objectContaining({ defaultSceneName: 'testScene' }),
      expect.anything()
    );
    expect(getByText('Test Child Component')).toBeTruthy();
  });

  test('should not render the SceneProvider if the scene is not active', () => {
    // Set the mock to return inactive state
    mockUseSceneContext.mockReturnValueOnce({ isActive: false });

    const { queryByText } = render(<RenderWithContext />);

    expect(SceneProvider).not.toHaveBeenCalled();
    expect(queryByText('Test Child Component')).toBeNull();
  });

  test('should correctly render nested scenes when both are active', () => {
    mockUseSceneContext.mockReturnValue({ isActive: true });

    const { getByText } = render(
      <Scene defaultSceneName="parentScene">
        <Scene defaultSceneName="nestedScene">
          <Text>Nested Scene Content</Text>
        </Scene>
        <Text>Parent Scene Content</Text>
      </Scene>
    );

    expect(getByText('Parent Scene Content')).toBeTruthy();
    expect(getByText('Nested Scene Content')).toBeTruthy();
  });

  test('should not render nested scene if parent is inactive', () => {
    mockUseSceneContext.mockReturnValueOnce({ isActive: false }); // Mock parent inactive

    const { queryByText } = render(
      <Scene defaultSceneName="parentScene">
        <Scene defaultSceneName="nestedScene">
          <Text>Nested Scene Content</Text>
        </Scene>
        <Text>Parent Scene Content</Text>
      </Scene>
    );

    expect(queryByText('Parent Scene Content')).toBeNull();
    expect(queryByText('Nested Scene Content')).toBeNull(); // Should not render since parent is inactive
  });

  test('should update rendering when scene becomes active', () => {
    mockUseSceneContext.mockReturnValueOnce({ isActive: false });

    const { rerender, queryByText, getByText } = render(<RenderWithContext />);

    // Initially inactive
    expect(queryByText('Test Child Component')).toBeNull();

    // Become active
    mockUseSceneContext.mockReturnValueOnce({ isActive: true });
    rerender(<RenderWithContext />);
    expect(getByText('Test Child Component')).toBeTruthy();
  });

  test('should render complex nested scenes with dynamic states', () => {
    mockUseSceneContext
      .mockReturnValueOnce({ isActive: true }) // Parent active
      .mockReturnValueOnce({ isActive: false }); // Nested inactive

    const { getByText, queryByText } = render(
      <Scene defaultSceneName="parentScene">
        <Text>Parent Scene Content</Text>
        <Scene defaultSceneName="nestedScene">
          <Text>Nested Scene Content</Text>
          <Scene defaultSceneName="deeplyNestedScene">
            <Text>Deeply Nested Scene Content</Text>
          </Scene>
        </Scene>
      </Scene>
    );

    expect(getByText('Parent Scene Content')).toBeTruthy();
    expect(queryByText('Nested Scene Content')).toBeNull(); // Nested is inactive
    expect(queryByText('Deeply Nested Scene Content')).toBeNull(); // Deeply nested should not render
  });

  test('should match the snapshot for a nested active scene', () => {
    mockUseSceneContext.mockReturnValue({ isActive: true });

    const { toJSON } = render(
      <Scene defaultSceneName="parentScene">
        <Scene defaultSceneName="nestedScene">
          <Text>Nested Scene Content</Text>
        </Scene>
        <Text>Parent Scene Content</Text>
      </Scene>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  test('should match the snapshot for a nested inactive scene', () => {
    mockUseSceneContext.mockReturnValueOnce({ isActive: false });

    const { toJSON } = render(
      <Scene defaultSceneName="parentScene">
        <Scene defaultSceneName="nestedScene">
          <Text>Nested Scene Content</Text>
        </Scene>
        <Text>Parent Scene Content</Text>
      </Scene>
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
