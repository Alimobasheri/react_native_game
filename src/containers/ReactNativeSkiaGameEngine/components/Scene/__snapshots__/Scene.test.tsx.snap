// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scene should match the snapshot for a nested active scene 1`] = `
<skGroup
  clip={
    {
      "height": 300,
      "width": 300,
      "x": 0,
      "y": 0,
    }
  }
  opacity={
    {
      "_isReanimatedSharedValue": true,
      "_value": 1,
      "addListener": [Function],
      "modify": [Function],
      "removeListener": [Function],
      "value": 1,
    }
  }
  transform={
    {
      "_isReanimatedSharedValue": true,
      "_value": [
        {
          "translateY": 0,
        },
      ],
      "addListener": [Function],
      "modify": [Function],
      "removeListener": [Function],
      "value": [
        {
          "translateY": 0,
        },
      ],
    }
  }
>
  <skGroup
    clip={
      {
        "height": 300,
        "width": 300,
        "x": 0,
        "y": 0,
      }
    }
    opacity={
      {
        "_isReanimatedSharedValue": true,
        "_value": 1,
        "addListener": [Function],
        "modify": [Function],
        "removeListener": [Function],
        "value": 1,
      }
    }
    transform={
      {
        "_isReanimatedSharedValue": true,
        "_value": [
          {
            "translateY": 0,
          },
        ],
        "addListener": [Function],
        "modify": [Function],
        "removeListener": [Function],
        "value": [
          {
            "translateY": 0,
          },
        ],
      }
    }
  >
    <Text>
      Nested Scene Content
    </Text>
  </skGroup>
  <Text>
    Parent Scene Content
  </Text>
</skGroup>
`;

exports[`Scene should match the snapshot for a nested inactive scene 1`] = `null`;
